Add GitHub Actions workflow to build the `stitching` C++ binary

joshuacwnewton opened the pull request:
This PR contains @jqmcginnis's previous debugging work to develop a workflow for building the `stitching` C++ binary.

> **NB:** I've marked this as draft as there's still some work in progress to get it working for Linux/macOS/Windows. :slightly_smiling_face:

jqmcginnis committed:
starting with github workflows

jqmcginnis committed:
create first github action

jqmcginnis committed:
update workflow yaml

jqmcginnis committed:
bugfix

jqmcginnis committed:
bugfix (master and main)

jqmcginnis committed:
add reference

jqmcginnis committed:
update

jqmcginnis committed:
bugfix

jqmcginnis committed:
add debugging

jqmcginnis committed:
small change

jqmcginnis committed:
bugfix

jqmcginnis committed:
add complete stitching build workflow

jqmcginnis committed:
test workaround due to permission problem

jqmcginnis committed:
bugfix

jqmcginnis committed:
Update build-stitching.yml

jqmcginnis committed:
Update build-stitching.yml

jqmcginnis committed:
add debugging

joshuacwnewton commented on .github/workflows/build-stitching.yml:
```diff
@@ -0,0 +1,66 @@
+        make -j$(nproc)
+        make install
+
+    - name: get_boost_apt  
+      run: sudo apt-get install -y libboost-all-dev libtbb-dev
```
I've been trying to build `stitching` locally without using the prebuilt `Boost` apt package, and arrived at these steps:

```bash
cd $THIRD_PARTY_DIR
wget https://boostorg.jfrog.io/artifactory/main/release/1.80.0/source/boost_1_80_0.tar.gz --progress=bar:force:noscroll
tar xf boost_1_80_0.tar.gz 
cd boost_1_80_0/
mkdir build
# build Boost 
./bootstrap.sh --prefix=build/
./b2
# point cmake at the location of the various boost cmake files
export CMAKE_PREFIX_PATH=../3rdParty/boost_1_80_0/stage/lib/cmake/
```

I also had to point `stitching` at the built ITK libraries, too?

```bash
export ITK_DIR=../3rdParty/InsightToolkit-5.0.0/build
```

But apart from those steps, I was able to build `stitching` using the static Boost libraries just fine! 

```console
joshua@tadpole:~/Desktop/build_biomedia_stitching/build$ make
[ 12%] Building CXX object mia/CMakeFiles/mia.dir/src/miaHistogram.cpp.o
[ 25%] Building CXX object mia/CMakeFiles/mia.dir/src/miaImage.cpp.o
[ 37%] Building CXX object mia/CMakeFiles/mia.dir/src/miaImageProcessing.cpp.o
[ 50%] Linking CXX static library libmia.a
[ 50%] Built target mia
[ 62%] Building CXX object itkio/CMakeFiles/itkio.dir/src/itkio.cpp.o
[ 75%] Linking CXX static library libitkio.a
[ 75%] Built target itkio
[ 87%] Building CXX object stitching/CMakeFiles/stitching.dir/src/main.cpp.o
[100%] Linking CXX executable stitching
[100%] Built target stitching
```

I'll see if I can port these steps to our GitHub Actions workflow now. :slightly_smiling_face:

joshuacwnewton replied:
I then tried running `sct_image -stitch` command from https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3865#discussion_r956085087 using this new binary, and it looks like it worked for me?

```console
joshua@tadpole:~/Desktop/build_biomedia_stitching$ cp build/stitching/stitching $SCT_DIR/bin
joshua@tadpole:~/Desktop/build_biomedia_stitching$ cd $SCT_DIR/data/sct_testing_data/t2
joshua@tadpole:~/repos/spinalcordtoolbox/data/sct_testing_data/t2$ sct_image -i sct_roi1.nii.gz sct_roi2.nii.gz -o stitched.nii.gz -stitch

--
Spinal Cord Toolbox (git-remotes/origin/jqm/3828-sct-stitching-dfb91f755620dbeca936831d2a451848c6ddd753)

sct_image -i sct_roi1.nii.gz sct_roi2.nii.gz -o stitched.nii.gz -stitch
--

/tmp/sct-20220826135409.913820-image-stitching-pg8vv26f/sct_roi1.nii.gz
cp sct_roi1.nii.gz /tmp/sct-20220826135409.913820-image-stitching-pg8vv26f/sct_roi1.nii.gz
File /tmp/sct-20220826135409.913820-image-stitching-pg8vv26f/sct_roi1.nii.gz already exists. Will overwrite it.
/tmp/sct-20220826135409.913820-image-stitching-pg8vv26f/sct_roi2.nii.gz
cp sct_roi2.nii.gz /tmp/sct-20220826135409.913820-image-stitching-pg8vv26f/sct_roi2.nii.gz
File /tmp/sct-20220826135409.913820-image-stitching-pg8vv26f/sct_roi2.nii.gz already exists. Will overwrite it.
/home/joshua/repos/spinalcordtoolbox/bin/stitching -i 'sct_roi1.nii.gz sct_roi2.nii.gz' -o /tmp/sct-20220826135409.913820-image-stitching-pg8vv26f/stitched.nii.gz -a # in /tmp/sct-20220826135409.913820-image-stitching-pg8vv26f
0 stitching image...done. took 70 ms
rm -rf /tmp/sct-20220826135409.913820-image-stitching-pg8vv26f
Generate output files...

Done! To view results, run one of the following commands (depending on your preferred viewer):
fsleyes stitched.nii.gz &

fslview_deprecated stitched.nii.gz &

fslview stitched.nii.gz &
```

Just as a sanity check (to make sure I didn't accidentally install `boost` systemwide), I switched back to the old `stitching` binary, and got this:

```console
(venv_sct) joshua@tadpole:~/repos/spinalcordtoolbox/data/sct_testing_data/t2$ sct_image -i sct_roi1.nii.gz sct_roi2.nii.gz -o stitched.nii.gz -stitch

--
Spinal Cord Toolbox (git-remotes/origin/jqm/3828-sct-stitching-dfb91f755620dbeca936831d2a451848c6ddd753)

sct_image -i sct_roi1.nii.gz sct_roi2.nii.gz -o stitched.nii.gz -stitch
--

/tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz/sct_roi1.nii.gz
cp sct_roi1.nii.gz /tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz/sct_roi1.nii.gz
File /tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz/sct_roi1.nii.gz already exists. Will overwrite it.
/tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz/sct_roi2.nii.gz
cp sct_roi2.nii.gz /tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz/sct_roi2.nii.gz
File /tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz/sct_roi2.nii.gz already exists. Will overwrite it.
/home/joshua/repos/spinalcordtoolbox/bin/stitching -i 'sct_roi1.nii.gz sct_roi2.nii.gz' -o /tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz/stitched.nii.gz -a # in /tmp/sct-20220826135922.332441-image-stitching-f8sq2tqz
Traceback (most recent call last):
  File "/home/joshua/repos/spinalcordtoolbox/spinalcordtoolbox/scripts/sct_image.py", line 607, in <module>
    main(sys.argv[1:])
  File "/home/joshua/repos/spinalcordtoolbox/spinalcordtoolbox/scripts/sct_image.py", line 361, in main
    im_out = stitch_images(fnames_in=fnames_in)
  File "/home/joshua/repos/spinalcordtoolbox/spinalcordtoolbox/image.py", line 1724, in stitch_images
    status, output = run_proc(cmd, verbose='verbose', is_sct_binary=True)
  File "/home/joshua/repos/spinalcordtoolbox/spinalcordtoolbox/utils/sys.py", line 414, in run_proc
    raise RuntimeError(output)
RuntimeError: /home/joshua/repos/spinalcordtoolbox/bin/stitching: error while loading shared libraries: libboost_program_options.so.1.71.0: cannot open shared object file: No such file or directory
```

:D

jqmcginnis replied:
Excellent work @joshuacwnewton ! :+1:

jqmcginnis replied:
@joshuacwnewton do you want to push the changes :) ?

joshuacwnewton replied:
I'm happy to take care of this, yes! :)

joshuacwnewton committed:
`build-stitching.yml`: Use `boost` source archive, not `apt`

joshuacwnewton committed:
`build-stitching.yml`: Merge workflow steps into single step

This is just temporary, and will save time in debugging. This is because environment variables don't usually carry over between steps. So, we won't have to do any sort of `echo "export THIRD_PARTY_DIRECTORY=3rdParty" >> $GITHUB_ENV` commands if we keep it all in one step for now.

See also: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable

joshuacwnewton committed:
`build-stitching.yml`: Increase timeout

The build step timed out on the previous run, because all of the steps are now in 1 build.

joshuacwnewton committed:
`build-stitching.yml`: Remove `ls` debugging commands

joshuacwnewton committed:
`build-stitching.yml`: Comment out working build steps to save time

This is just to debug the "cd build" in the stitching step.

joshuacwnewton committed:
`build-stitching.yml`: Try to fixup `cmake syntax`

joshuacwnewton committed:
`build-stitching.yml`: Fix comment indentation

If comments aren't indented correctly, the .yml won't be valid.

joshuacwnewton committed:
`build-stitching.yml`: Uncomment main steps

Now that I've confirmed that the cmake syntax fix works, I can run the main steps

joshuacwnewton committed:
`build-stitching.yml`: Use workflow-wide $THIRD_PARTY_DIR

joshuacwnewton committed:
`build-stitching.yml`: Inline the VERSION variables

While this works OK for the specified versions, there's no guarantee these URL schemes will work for future versions, so IMO it's a little more straightforward just to hardcode the version strings in the URLs

joshuacwnewton committed:
`build-stitching.yml`: Break workflow into multiple steps

joshuacwnewton committed:
`build-stitching.yml`: Ensure static linking for Boost libraries

joshuacwnewton committed:
`build-stitching.yml`: Add dynamic linking check

joshuacwnewton committed:
`build-stitching.yml`: Add packaging and upload step

joshuacwnewton committed:
`build-stitching.yml`: Fixup environment variables for stitching build

joshuacwnewton committed:
`build-stitching.yml`: Change workflow name from 'Run tests'

joshuacwnewton committed:
`build-stitching.yml`: Remove redundant "push" workflow trigger

joshuacwnewton committed:
`build-stitching.yml`: Downgrade runner OS to `ubuntu-18.04`

joshuacwnewton committed:
`build-stitching.yml`: Try building against both `ubuntu-18.04` and `macos-10.15`

joshuacwnewton commented on .github/workflows/build-stitching.yml:
```diff
@@ -0,0 +1,92 @@
+jobs:
+  build:
+    strategy:
+      matrix:
+        os: [ubuntu-18.04, macos-10.15]
```
I've tried adding `macos-10.15` without changing anything else, and I've immediately run into a whole host of [errors/warnings](https://github.com/jqmcginnis/build_biomedia_stitching/runs/8082494426?check_suite_focus=true#step:7:300) associated with the [`mia`](https://github.com/biomedia-mira/stitching/tree/master/mia) library distributed alongside `stitching`.

Many of the warnings reference C++11 extensions, so it might be worth following the advice of [this SO answer](https://stackoverflow.com/a/45291997) and compile with `-std=c++11` if possible. 

(Alternatively, specifying C++11 through CMake might be done via [these instructions instead](https://stackoverflow.com/questions/10851247/how-do-i-activate-c-11-in-cmake). Additionally. see [this SO post](https://stackoverflow.com/questions/24589105/what-is-the-difference-between-std-c0x-and-std-c11) for clarification between `c++0x` and `c++11`.)

joshuacwnewton replied:
Ah, now I'm definitely sure this is related to the compiler not using the C++11 standard. One of the errors is:

```
/Users/runner/work/build_biomedia_stitching/build_biomedia_stitching/mia/src/miaImage.cpp:68:21: error: expected expression
    , m_data( data, [](float*) {} )
                    ^
```

This `[](float*)` syntax is a [lambda expression](https://docs.microsoft.com/en-us/cpp/cpp/lambda-expressions-in-cpp?view=msvc-170), a feature that was [introduced in C++11](https://smartbear.com/blog/the-biggest-changes-in-c11-and-why-you-should-care/).

joshuacwnewton replied:
Good news! Adding `set(CMAKE_CXX_STANDARD 11)` allowed `mia` to build error-free on macOS. 

(The [error](https://github.com/jqmcginnis/build_biomedia_stitching/runs/8134881601?check_suite_focus=true) is due to `ldd` not being present on macOS, and has nothing to do with the actual build process.)

So, all that's left is Windows. :smile:

jqmcginnis replied:
@joshuacwnewton awesome news!

joshuacwnewton committed:
`build-stitching.yml`: Add setting to force C++11 for `mia` library

joshuacwnewton committed:
`build-stitching.yml`: Adapt ldd tool for macOS

joshuacwnewton committed:
`build-stitching.yml`: Add first draft Windows build job

joshuacwnewton committed:
`build-stitching.yml`: Temporarily disable unix builds to save time/resources

joshuacwnewton committed:
`build-stitching.yml`: Fix typo in `curl` command

joshuacwnewton committed:
`build-stitching.yml`: Add `msbuild-architecture` setting

The workflow mentions "msbuild: command not found" but this is the exact step used in `build_ANTs.yml`, so I'm not really sure what's going on here.

joshuacwnewton committed:
`build-stitching.yml`: Split setup and build steps

This allows us to specifically use `shell: cmd` for the build steps

joshuacwnewton committed:
`build-stitching.yml`: Replace `cd` with `working_directory:`

joshuacwnewton committed:
`build-stitching.yml`: Stop using `$GITHUB_WORKSPACE` in THIRD_PARTY_DIR

joshuacwnewton committed:
`build-stitching.yml`: Use relative paths for stitching environment variables

Since THIRD_PARTY_DIR no longer includes the absolute path, we have to specify relative paths for these environment variables.

joshuacwnewton committed:
`build-stitching.yml`: Re-enable `build-unix`

joshuacwnewton committed:
`build-stitching.yml`: Add release-creation step

joshuacwnewton committed:
`build-stitching.yml`: Use relative paths for unix, too
